### PR TITLE
Implement roles & permissions

### PR DESCRIPTION
**Pull request by yana617**
Blocked by PR #1

### CODE REVIEW

File: src/routes/users.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    res.json({ success: true, data: user });
+  } catch (e) {
+    res.status(500).json({ success: false, error: e.message });
+  }
+});
+
+route.get('/:id', verifyToken, checkPermissions(['VIEW_USERS']), async (req, res) => {
+  try {
+    const { id: userId } = req.params;
+    const user = await models.User.findByPk(userId, { attributes: { exclude: ['hash', 'salt'] }, raw: true });
```
Comment (yana617) [inline src/routes/users.js:24]: include role

File: src/controllers/permission.controller.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    const mappedPermissionsToUpdate = permissionsToUpdate
+      .map((p) => ({ ...p, value: permissions[p.name] }));
+
+    const rolePerms = await rolePermissionRepository.getByRoleId(user.role_id);
+    const rolePermissionsIds = rolePerms.map((p) => p.permission_id);
+
+    const userPermissions = await userPermissionRepository.getByUserId(user.id);
+    const userPermissionsIds = userPermissions.map((p) => p.permission_id);
+
+    await Promise.all(mappedPermissionsToUpdate.map((permission) => {
```
Comment (yana617) [inline src/controllers/permission.controller.js:45]: fail if any role contains those permissions

**Comment by yana617** [comment]
implement translates

File: src/migrations/20210810130740-create-role-permissions.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: async (queryInterface, Sequelize) => {
+    await queryInterface.dropTable('role_permissions');
+  }
+};
```
Comment (Aphelmion) [inline src/migrations/20210810130740-create-role-permissions.js:24]: empty line

File: src/migrations/20210811203939-create-permission.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: async (queryInterface, Sequelize) => {
+    await queryInterface.dropTable('permissions');
+  }
+};
```
Comment (Aphelmion) [inline src/migrations/20210811203939-create-permission.js:29]: line

File: src/migrations/20210811204343-create-role.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: async (queryInterface, Sequelize) => {
+    await queryInterface.dropTable('roles');
+  }
+};
```
Comment (Aphelmion) [inline src/migrations/20210811204343-create-role.js:32]: line

File: src/services/permissions.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  rolePermissionsIds,
+  userRoleId,
+) => {
+  const allRolesWithPermissions = await roleRepository.getAllWithPermissions();
+  const mappedRoles = allRolesWithPermissions.map((r) => ({
+    name: r.name,
+    permissionsIds: r.role_permissions.map((rp) => rp.permission_id),
+  }));
+  const biggerRoles = mappedRoles.filter((r) => !r.permissionsIds
+    .every((v) => rolePermissionsIds.includes(v)) && r.id !== userRoleId);
```
Comment (Aphelmion) [inline src/services/permissions.js:42]: permissionId

File: src/services/permissions.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  userPermissionsIds,
+  rolePermissionsIds,
+  userRoleId,
+) => {
+  const allRolesWithPermissions = await roleRepository.getAllWithPermissions();
+  const mappedRoles = allRolesWithPermissions.map((r) => ({
+    name: r.name,
+    permissionsIds: r.role_permissions.map((rp) => rp.permission_id),
+  }));
+  const biggerRoles = mappedRoles.filter((r) => !r.permissionsIds
```
Comment (Aphelmion) [inline src/services/permissions.js:41]: rolesSeniorThanUserRole

**Review by Aphelmion — APPROVED** [review]

**Pull request merged**